spec/lib/msf/base/serializer/readable_text_spec.rb

```diff
@@ -0,0 +1,158 @@
+# -*- coding:binary -*-
+
+require 'spec_helper'
+require 'rex/text'
+
+RSpec.describe Msf::Serializer::ReadableText do
+  # The described_class API takes a mix of strings and whitespace character counts
+  let(:indent_string) { '' }
+  let(:indent_length) { indent_string.length }
+
+  let(:aux_mod) do
+    mod_klass = Class.new(Msf::Auxiliary) do
+      def initialize
+        super(
+          'Name' => 'mock module',
+          'Description' => 'mock module',
+          'Author' => ['Unknown'],
+          'License' => MSF_LICENSE,
+          'DefaultOptions' => {
+            'OptionWithModuleDefault' => false,
+            'foo' => 'foo_from_module',
+            'baz' => 'baz_from_module'
+          },
+        )
+
+        register_options(
+          [
+            Msf::Opt::RHOSTS,
+            Msf::Opt::RPORT(3000),
+            Msf::OptString.new(
+              'foo',
+              [true, 'Foo option', 'bar']
+            ),
+            Msf::OptString.new(
+              'fizz',
+              [true, 'fizz option', 'buzz']
+            ),
+            Msf::OptString.new(
+              'baz',
+              [true, 'baz option', 'qux']
+            ),
+            Msf::OptString.new(
+              'OptionWithModuleDefault',
+              [true, 'option with module default', true]
+            ),
+            Msf::OptFloat.new('FloatValue', [false, 'A FloatValue ', 3.5]),
+            Msf::OptString.new(
+              'NewOptionName',
+              [true, 'An option with a new name. Aliases ensure the old and new names are synchronized', 'default_value'],
+              aliases: ['OLD_OPTION_NAME']
+            ),
+            Msf::OptString.new(
+              'SMBUser',
+              [true, 'The SMB username'],
+              fallbacks: ['username']
+            ),
+            Msf::OptString.new(
+              'SMBDomain',
+              [true, 'The SMB username', 'WORKGROUP'],
+              fallbacks: ['domain']
+            )
+          ]
```

Comment:
What is the expected result for:
```
           Msf::OptString.new(
              'SMBDomain',
              [true, 'The SMB username', 'WORKGROUP'],
              alias: ['WindowsDomain']
              fallbacks: ['domain']
            )
```

A contrived scenario but worth documenting expectation in spec.

Reply:
I've added an extra test for this:

```
    context 'when the datastore has aliases and fallbacks' do
      subject do
        s = default_subject.copy
        options = Msf::OptionContainer.new(
          [
            Msf::OptString.new(
              'SMBDomain',
              [true, 'The SMB username', 'WORKGROUP'],
              aliases: ['WindowsDomain'],
              fallbacks: ['domain']
            )
          ]
        )
        s.import_options(options)
        s
      end

      context 'when the fallback value is set' do
        before(:each) do
          subject['domain'] = 'domain_fallback'
        end

        it 'supports reading with the option name' do
          expect(subject['SMBDomain']).to eq('domain_fallback')
        end

        it 'supports reading with the alias name' do
          expect(subject['WindowsDomain']).to eq('domain_fallback')
        end
      end
```

Let me know if there was a different testing scenario you were after :+1:

Reply:
I've also added `aliases: ['WindowsDomain']` to this file; By default adding that line to the option won't do anything as `WindowsDomain` hasn't been set, so the test's expectations did not have to be modified

Reply:
One more test that defines that `alias` is higher priority than a fallback may be helpful.

``` Ruby
     context 'when the alias and fallback value are set' do 
       before(:each) do
          subject['domain'] = 'domain_fallback'
          subject['WindowsDomain'] = 'WindowsDomain'
        end

        it 'supports reading with the option name' do
          expect(subject['SMBDomain']).to eq('WindowsDomain')
        end

        it 'supports reading with the alias name' do
          expect(subject['SMBDomain']).to eq('WindowsDomain')
        end
      end
```

Reply:
Test added, no changes were required in the implementation :+1: